### ISSUE TITLE
Unnecessary to create ~/.zshenv when dotDir is set

### ISSUE THREAD

**Issue by Th3Whit3Wolf** [nix-community/home-manager#1965]
### Issue description

Home manager creates a `~/.zshenv` unnecessarily, cluttering the home directory.

https://github.com/nix-community/home-manager/blob/a759143ae130e68b8c0ce867c28b3e006f9920e0/modules/programs/zsh.nix#L407-L417

it would be nice to either have an option to not create this file or instead make this global like this.

```nix
environment.etc."zsh/zshenv" = {
  text = ''
    ZDOTDIR=${zdotdir}
  '';
};
```

**Comment by berbiche** [comment]
Home Manager only changes files in a user's home directory.
Changing things globally is outside the scope of HM and will probably never be implemented.

Technically, you can probably make a dummy configuration where the `homeDirectory` is `/etc`, use an empty configuration and only write files using `home.files` (e.g. `home.files."zsh/zshenv".text = "ZDOTDIR=something"`).

**Comment by Th3Whit3Wolf** [comment]
> Home Manager only changes files in a user's home directory.
Changing things globally is outside the scope of HM and will probably never be implemented.

That's what I figured, I put the snippet their really for anyone who wasn't familiar with the solution.

> Technically, you can probably make a dummy configuration where the homeDirectory is /etc, use an empty configuration and only write files using home.files (e.g. home.files."zsh/zshenv".text = "ZDOTDIR=something").

I think I'd prefer be able to set a boolean to not create `~/.zshenv`

**Comment by Th3Whit3Wolf** [comment]
I think this maybe possible using `systemd.user.sessionVariables`

**Comment by fricklerhandwerk** [comment]
> Home manager creates a ~/.zshenv unnecessarily, cluttering the home directory.

It only happens if `zsh` itself is enabled, and even then only if you

- add `oh-my-zsh`

https://github.com/nix-community/home-manager/blob/a759143ae130e68b8c0ce867c28b3e006f9920e0/modules/programs/zsh.nix#L399-L404

- have extra lines to add to the environment

https://github.com/nix-community/home-manager/blob/a759143ae130e68b8c0ce867c28b3e006f9920e0/modules/programs/zsh.nix#L382-L385

- relocate the config directory

https://github.com/nix-community/home-manager/blob/a759143ae130e68b8c0ce867c28b3e006f9920e0/modules/programs/zsh.nix#L406-L418

Running `rg zshenv`, `rg oh-my-zsh`, `rg envExtra` on the repo does not show any other relevant results.

Did you enable any of the above?

**Comment by Th3Whit3Wolf** [comment]
@fricklerhandwerk yes I use an XDG Compliant directory for my configs. I have renamed the issue to reflect that.

**Comment by uvNikita** [comment]
I might be missing something here, but as far as I remember `~/.zshenv` is necessary to set ZDOTDIR environment variable in the first place, otherwise how zsh will know where config files are located?

**Comment by Th3Whit3Wolf** [comment]
@uvNikita when I get some free time I'm going to try to set it with systemd user environment variables. I could be wrong but I believe they are set by logind and before zsh

**Comment by teto** [comment]
@uvNikita wouldn't it be possible via sessionVariables ?

**Comment by uvNikita** [comment]
@teto I don't think so, because, unless I'm missing something, those variables are set via `zshrc`:

https://github.com/nix-community/home-manager/blob/f9e45390deced72bc7cc0b75d3a922aa10a33e08/modules/programs/zsh.nix#L480

**Comment by Th3Whit3Wolf** [comment]
@teto @uvNikita hm-sessiom-vars.sh contains variables set by `home. sessionVariablesExtra` I think  `systemd.user.sessionVariables` might work

**Comment by uvNikita** [comment]
@Th3Whit3Wolf I haven't tried it, but seems like `environment.d` (which is generated from `systemd.user.sessionVariables`) is used to set environment for services only(?):

> Configuration files in the environment.d/ directories contain lists of environment variable assignments for services started by the systemd user instance.

https://www.freedesktop.org/software/systemd/man/environment.d.html#Description

So it might work for graphical sessions if they would be started as systemd services, but I don't think it would work for something like ssh sessions.

**Comment by Th3Whit3Wolf** [comment]
I haven't tried it either. But I know it is the recommended way to set user environment variables under wayland.

[gnome recommends it here](https://wiki.gnome.org/Initiatives/Wayland/SessionStart)

**Comment by rycee** [comment]
I think it may be worth exploring use of environment.d for possible introduction in the next release cycle but it has to be done with care, we'd want to avoid problems like in https://github.com/nix-community/home-manager/issues/1999.

**Comment by fricklerhandwerk** [comment]
Yes, and let's not forget that `systemd` is not the only mechanism out there. We have a substantial user base on Darwin, for which #1999 also [applies](https://github.com/nix-community/home-manager/issues/1999#issuecomment-843858324).

**Comment by utkarshgupta137** [comment]
Not completed

**Comment by Theaninova** [comment]
For anyone interested, you can work around this right now (in my case I have hm as a NixOS module):
`clean-zsh.nix`
```nix
{ config, ... }:
let
  xdgConfig = config.home-manager.users.YOUR_USERNAME.xdg
in
{
  environment.etc."zshenv".text = ''
    source ${xdgConfig.configHome}/zsh/.zshenv
  '';
  home-manager.users.YOUR_USERNAME = {
    home.file.".zshenv".enable = false;
    zsh = {
      dotDir = ".config/zsh";
      history.path = "${xdgConfig.stateHome}/zsh/history";
    };
  };
}
```
`configuration.nix`
```nix
{ ... }:
{
  imports = [ ./clean-zsh.nix ];
}
```
*I got a more elaborate version of this [here](https://github.com/Theaninova/TheaninovOS/blob/5196b2bab7abc60aea033b60a19484d8924f2fc1/modules/nixos/xdg/forced-compliance.nix#L95)*
